### PR TITLE
Clean FID eval for SDXL

### PR DESCRIPTION
**Pull request by jazcollins**
This PR makes two changes:

1. Small fixes necessary to get make FID eval compatible with SDXL
2. Modify `generate()` function to allow passing SDXL micro-conditioning as either a tuple (one set of crop/size parameters that will replicated across the batch dimension) or a tensor (batch of different crop/size parameters) - the latter of which is necessary for 1.

### CODE REVIEW

File: diffusion/evaluation/clean_fid_eval.py [hunk 1]
```diff
@@ -62,6 +63,7 @@ def __init__(self,
                  batch_size: int = 16,
                  image_key: str = 'image',
                  caption_key: str = 'caption',
+                 load_strict_model_weights: bool = True,
```
Comment (Skylion007) [inline diffusion/evaluation/clean_fid_eval.py:66]: Isn't this already specified by the "trainer"?  Can we reuse those settings?

File: diffusion/models/stable_diffusion.py [hunk 2]
```diff
@@ -342,8 +342,8 @@ def generate(
         seed: Optional[int] = None,
         progress_bar: Optional[bool] = True,
         zero_out_negative_prompt: bool = True,
-        crop_params: Optional[list] = None,
-        size_params: Optional[list] = None,
+        crop_params: Optional[Union[tuple, torch.Tensor]] = None,
```
Comment (Skylion007) [inline diffusion/models/stable_diffusion.py:345]: Tuple should of what typing / what size?

File: diffusion/models/stable_diffusion.py [hunk 3]
```diff
@@ -448,14 +452,24 @@ def generate(
...
+            if crop_params is None:
+                crop_params = torch.tensor([0., 0.]).repeat(batch_size, 1)
+            elif isinstance(crop_params, tuple):  # convert to tensor
+                crop_params = torch.tensor([crop_params[0], crop_params[1]]).repeat(batch_size, 1).float()
+            if input_size_params is None:
+                input_size_params = torch.tensor([width, height]).repeat(batch_size, 1).float()
+            elif isinstance(input_size_params, tuple):  # convert to tensor
+                input_size_params = torch.tensor([input_size_params[0], input_size_params[1]]).repeat(batch_size,
+                                                                                                      1).float()
+            output_size_params = torch.tensor([width, height]).repeat(batch_size, 1).float()
```
Comment (Skylion007) [inline diffusion/models/stable_diffusion.py:464]: Just construct the tensor with dtype float?

File: diffusion/evaluation/clean_fid_eval.py (hunk 1 above)
Comment (jazcollins) [reply diffusion/evaluation/clean_fid_eval.py:66]: hm can you elaborate on how to do this?

File: diffusion/models/stable_diffusion.py (hunk 3 above)
Comment (jazcollins) [reply diffusion/models/stable_diffusion.py]: done!

File: diffusion/models/stable_diffusion.py (hunk 2 above)
Comment (jazcollins) [reply diffusion/models/stable_diffusion.py]: done!

File: diffusion/evaluation/clean_fid_eval.py [hunk 4]
```diff
@@ -95,6 +99,7 @@ def __init__(self,
         Trainer(model=self.model,
                 load_path=self.load_path,
                 load_weights_only=True,
+                load_strict_model_weights=self.load_strict_model_weights,
```
Comment (Skylion007) [inline diffusion/evaluation/clean_fid_eval.py:102]: Oh, I just meant forwarding all remaining kwargs to the trainer?

File: diffusion/evaluation/clean_fid_eval.py [hunk 5]
```diff
@@ -78,12 +80,14 @@ def __init__(self,
         self.batch_size = batch_size
         self.image_key = image_key
         self.caption_key = caption_key
+        self.load_strict_model_weights = load_strict_model_weights
```
Comment (Skylion007) [inline diffusion/evaluation/clean_fid_eval.py:83]: Can't we access this boolean from the trainer config later? Why bother storing a copy of it in the class?

**Review by Skylion007 — APPROVED** [review]

File: diffusion/models/stable_diffusion.py [hunk 6]
```diff
@@ -448,14 +452,24 @@ def generate(
...
         if self.sdxl and pooled_embeddings is not None:
-            if not crop_params:
-                crop_params = [0., 0.]
-            if not size_params:
-                size_params = [width, height]
-            add_time_ids = torch.tensor([[width, height, *crop_params, *size_params]], dtype=torch.float, device=device)
-            add_time_ids = add_time_ids.repeat(pooled_embeddings.shape[0], 1)
-            add_text_embeds = pooled_embeddings
+            if crop_params is None:
+                crop_params = torch.tensor([0., 0.]).repeat(batch_size, 1)
```
Comment (Skylion007) [inline diffusion/models/stable_diffusion.py:456]: Use torch.zeros here?

File: diffusion/models/stable_diffusion.py [hunk 7]
```diff
@@ -448,14 +452,24 @@ def generate(
...
+            output_size_params = torch.tensor([width, height], dtype=torch.float32).repeat(batch_size, 1)
 
+            if do_classifier_free_guidance:
+                crop_params = torch.cat([crop_params, crop_params])
+                input_size_params = torch.cat([input_size_params, input_size_params])
+                output_size_params = torch.cat([output_size_params, output_size_params])
+
+            add_time_ids = torch.cat([input_size_params, crop_params, output_size_params], dim=1).to(device)
+            add_text_embeds = pooled_embeddings
             added_cond_kwargs = {'text_embeds': add_text_embeds, 'time_ids': add_time_ids}
```
Comment (Landanjs) [inline diffusion/models/stable_diffusion.py:473]: Any reason to not do this?
```suggestion
            added_cond_kwargs = {'text_embeds': pooled_embeddings, 'time_ids': add_time_ids}
```

**Review by Landanjs — APPROVED** [review]
LGTM! Minor comments

File: diffusion/models/stable_diffusion.py [hunk 8]
```diff
@@ -448,14 +452,24 @@ def generate(
...
-                size_params = [width, height]
-            add_time_ids = torch.tensor([[width, height, *crop_params, *size_params]], dtype=torch.float, device=device)
-            add_time_ids = add_time_ids.repeat(pooled_embeddings.shape[0], 1)
-            add_text_embeds = pooled_embeddings
+            if crop_params is None:
+                crop_params = torch.tensor([0., 0.]).repeat(batch_size, 1)
+            elif isinstance(crop_params, tuple):  # convert to tensor
+                crop_params = torch.tensor([crop_params[0], crop_params[1]], dtype=torch.float32).repeat(batch_size, 1)
+            if input_size_params is None:
+                input_size_params = torch.tensor([width, height], dtype=torch.float32).repeat(batch_size, 1)
```
[the remark this answers is not in the document]
Comment (jazcollins) [reply diffusion/models/stable_diffusion.py:460]: iirc it gets set appropriately later on the in unet call, but I can set it here for clarity!

File: diffusion/models/stable_diffusion.py [hunk 9]
```diff
@@ -342,8 +342,8 @@ def generate(
         seed: Optional[int] = None,
         progress_bar: Optional[bool] = True,
         zero_out_negative_prompt: bool = True,
-        crop_params: Optional[list] = None,
-        size_params: Optional[list] = None,
+        crop_params: Optional[Union[Tuple[int, int], torch.Tensor]] = None,
+        input_size_params: Optional[Union[Tuple[int, int], torch.Tensor]] = None,
```
[the remark this answers is not in the document]
Comment (jazcollins) [reply diffusion/models/stable_diffusion.py:346]: will do!

**Review by coryMosaicML — APPROVED** [review]

File: diffusion/models/stable_diffusion.py [hunk 10]
```diff
@@ -313,7 +313,7 @@ def update_metric(self, batch, outputs, metric):
             if self.sdxl:
                 # Decode captions with first tokenizer
                 captions = [
-                    self.tokenizer.tokenizer.decode(caption[0], skip_special_tokens=True)
+                    self.tokenizer.tokenizer.decode(caption[:, 0, :][0], skip_special_tokens=True)
```
[the remark this answers is not in the document]
Comment (jazcollins) [reply diffusion/models/stable_diffusion.py:316]: yep! `caption` is [B, 2, seq_len] for SDXL, where the middle dimension is the two tokenized outputs stacked together

**Pull request merged**